flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapSnapshotResources.java

```diff
@@ -50,7 +51,7 @@
     private final TypeSerializer<K> keySerializer;
     private final int totalKeyGroups;
 
-    private HeapSnapshotResources(
+    protected HeapSnapshotResources(
```

Comment:
Why is it protected? The class is final...

Could we use composition instead of inheritance? Inheritance is pretty much always a wrong choice if used for code deduplication, as it creates complex hierarchies that are later on loosely coupled. I think we can pass the method `snapshot(long, CheckpointOptions, StateSnapshotRestore)` instead of extending the class.